Some classic queues don't have consumers after node restart

@acogoluegnes this may or may not be a perf-test issue. To be investigated further next week.

Given a bunch of classic queues with publishers and consumers
when one of the nodes in the cluster is restarted
some queues no longer have consumers

This was originally reported in the context of a Kubernetes Operator and a full rolling restart. However, seems like it can be reproduced locally with just a single node restart.

Steps:
```
$ make start-cluster PLUGINS="rabbitmq_management"

$ perf-test -x 10 -y 10 -r 10 -c 10 -A 10 -q 10 -qq -qp 'qq-%d' -qpf 0 -qpt 9 -H amqp://guest:guest@localhost:5672,amqp://guest:guest@localhost:5673,amqp://guest:guest@localhost:5674

$ perf-test -x 10 -y 10 -r 10 -c 10 -A 10 -q 10 -qp 'cq-%d' -qpf 0 -qpt 9 --auto-delete false -f persistent -H amqp://guest:guest@localhost:5672,amqp://guest:guest@localhost:5673,amqp://guest:guest@localhost:5674

$ ./sbin/rabbitmqctl -n rabbit-1 stop_app; ./sbin/rabbitmqctl -n rabbit-1 start_app
```

Before restarting the node, the state of the queues looks like this:
<img width="950" alt="before" src="https://user-images.githubusercontent.com/9566114/164544610-4ae0943c-15d0-4224-a9cd-13cee2035117.png">

After the restart, it looks like this:
<img width="946" alt="after" src="https://user-images.githubusercontent.com/9566114/164544634-11a3c1d5-ab4a-4705-8872-3ea946353838.png">

Notes:
* some of the queues don't have consumers, it may or may not be a perf-test re-connection issue
* it seems to always work well for QQs; sometimes it takes a moment to "heal" - a queue or two will take longer to get a consumer but they do get it; meanwhile, some CQs just don't
* the issue doesn't occur in a single-node RabbitMQ (just replace `start-cluster` with `run-broker` and all is good)

In my tests, perf-test logs this error:
```
Caused by: com.rabbitmq.client.ShutdownSignalException: channel error;
  protocol method: #method<channel.close>(reply-code=404, reply-text=NOT_FOUND
  - queue 'cq-9' in vhost '/' process is stopped by supervisor, class-id=60, method-id=20)
```

The original issue report also contains this:
```
Caused by: com.rabbitmq.client.ShutdownSignalException: channel error;
  protocol method: #method<channel.close>(reply-code=404, reply-text=NOT_FOUND
  - failed to perform operation on queue 'perf-test-10q-rate100-5' in vhost 'perf-test-grid' due to timeout, class-id=60, method-id=20)
```

Originally reported on Slack: https://rabbitmq.slack.com/archives/CTMSV81HA/p1650538594525179
There are many additional details and potential issues. The above doesn't match exactly what was reported but is a simplified case of one of the issues that rules out mirroring/auto-delete/non-durable properties as the cause. The above uses durable, non-mirrored queues and still has the same behaviour of "losing consumers".

--- a/README.md
+++ b/README.md
@@ -79,7 +79,7 @@ You need a local running RabbitMQ instance.
 
 Start a RabbitMQ container:
 
-    docker run -it --rm --name rabbitmq -p 5672:5672 rabbitmq:3.8
+    docker run -it --rm --name rabbitmq -p 5672:5672 rabbitmq:3.10
 
 Run the test suite:
 

--- a/src/main/java/com/rabbitmq/perf/AgentBase.java
+++ b/src/main/java/com/rabbitmq/perf/AgentBase.java
@@ -30,6 +30,16 @@ public abstract class AgentBase {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(AgentBase.class);
 
+    private volatile TopologyRecording topologyRecording;
+
+    public void setTopologyRecording(TopologyRecording topologyRecording) {
+        this.topologyRecording = topologyRecording;
+    }
+
+    protected TopologyRecording topologyRecording() {
+        return this.topologyRecording;
+    }
+
     protected void delay(long now, AgentState state) {
 
         long elapsed = now - state.getLastStatsTime();

--- a/src/main/java/com/rabbitmq/perf/Consumer.java
+++ b/src/main/java/com/rabbitmq/perf/Consumer.java
@@ -1,4 +1,4 @@
-// Copyright (c) 2007-2020 VMware, Inc. or its affiliates.  All rights reserved.
+// Copyright (c) 2007-2022 VMware, Inc. or its affiliates.  All rights reserved.
 //
 // This software, the RabbitMQ Java client library, is triple-licensed under the
 // Mozilla Public License 2.0 ("MPL"), the GNU General Public License version 2
@@ -19,7 +19,11 @@ import com.rabbitmq.client.AMQP.BasicProperties;
 import com.rabbitmq.client.*;
 import com.rabbitmq.client.AMQP.Queue.DeclareOk;
 import com.rabbitmq.perf.PerfTest.EXIT_WHEN;
+import com.rabbitmq.perf.TopologyRecording.RecordedQueue;
 import java.time.Duration;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -85,6 +89,8 @@ public class Consumer extends AgentBase implements Runnable {
 
     private volatile long lastDeliveryTag, lastAckedDeliveryTag;
 
+    private final ScheduledExecutorService topologyRecoveryScheduledExecutorService;
+
     public Consumer(ConsumerParameters parameters) {
         this.channel           = parameters.getChannel();
         this.id                = parameters.getId();
@@ -101,6 +107,7 @@ public class Consumer extends AgentBase implements Runnable {
         this.pollingInterval   = parameters.getPollingInterval();
         this.consumerArguments = parameters.getConsumerArguments();
         this.exitWhen          = parameters.getExitWhen();
+        this.topologyRecoveryScheduledExecutorService = parameters.getTopologyRecoveryScheduledExecutorService();
 
         this.queueNames.set(new ArrayList<>(parameters.getQueueNames()));
         this.initialQueueNames = new ArrayList<>(parameters.getQueueNames());
@@ -313,8 +320,9 @@ public class Consumer extends AgentBase implements Runnable {
             System.out.printf("Consumer cancelled by broker for tag: %s", consumerTag);
             if (consumerTagBranchMap.containsKey(consumerTag)) {
                 String qName = consumerTagBranchMap.get(consumerTag);
-                System.out.printf("Re-consuming. Queue: %s for Tag: %s", qName, consumerTag);
-                channel.basicConsume(qName, autoAck, consumerArguments, q);
+                TopologyRecording topologyRecording = topologyRecording();
+                RecordedQueue queueRecord = topologyRecording.queue(qName);
+                consumeOrScheduleConsume(queueRecord, topologyRecording, consumerTag, qName);
             } else {
                 System.out.printf("Could not find queue for consumer tag: %s", consumerTag);
             }
@@ -348,10 +356,12 @@ public class Consumer extends AgentBase implements Runnable {
         } else {
             for (Map.Entry<String, String> entry : consumerTagBranchMap.entrySet()) {
                 String queueName = queueName(topologyRecording, entry.getValue());
+                String consumerTag = entry.getKey();
                 LOGGER.debug("Recovering consumer, starting consuming on {}", queueName);
                 try {
-                    channel.basicConsume(queueName, autoAck, entry.getKey(), false, false, this.consumerArguments, q);
-                } catch (IOException e) {
+                    TopologyRecording.RecordedQueue queueRecord = topologyRecording.queue(entry.getValue());
+                    consumeOrScheduleConsume(queueRecord, topologyRecording, consumerTag, queueName);
+                } catch (Exception e) {
                     LOGGER.warn(
                             "Error while recovering consumer {} on queue {} on connection {}",
                             entry.getKey(), queueName, channel.getConnection().getClientProvidedName(), e
@@ -359,6 +369,55 @@ public class Consumer extends AgentBase implements Runnable {
                 }
             }
         }
+    }
+
+    private void consumeOrScheduleConsume(RecordedQueue queueRecord,
+                                          TopologyRecording topologyRecording,
+                                          String consumerTag,
+                                          String queueName) throws IOException {
+        if (queueMayBeDown(queueRecord, topologyRecording)) {
+            // If the queue is on a cluster node that is down, basic.consume will fail with a 404.
+            // This will close the channel and we can't afford it, so we check if the queue exists with a different channel,
+            // and postpone the subscription if the queue does not exist
+            LOGGER.debug("Checking if queue {} exists before subscribing", queueName);
+            if (Utils.exists(channel.getConnection(), ch -> ch.queueDeclarePassive(queueName))) {
+                LOGGER.debug("Queue {} does exist, subscribing", queueName);
+                channel.basicConsume(queueName, autoAck, consumerTag, false, false, this.consumerArguments, q);
+            } else {
+                LOGGER.debug("Queue {} does not exist, it is likely unavailable, scheduling subscription.", queueName);
+                Duration schedulingPeriod = Duration.ofSeconds(5);
+                int maxRetry = (int) (Duration.ofMinutes(10).getSeconds() / schedulingPeriod.getSeconds());
+                AtomicInteger retryCount = new AtomicInteger(0);
+                AtomicReference<Callable<Void>> resubscriptionReference = new AtomicReference<>();
+                Callable<Void> resubscription = () -> {
+                    LOGGER.debug("Scheduled re-subscription for {}...", queueName);
+                    if (Utils.exists(channel.getConnection(), ch -> ch.queueDeclarePassive(queueName))) {
+                        LOGGER.debug("Queue {} exists, re-subscribing", queueName);
+                        channel.basicConsume(queueName, autoAck, consumerTag, false, false, consumerArguments, q);
+                    } else if (retryCount.incrementAndGet() <= maxRetry){
+                        LOGGER.debug("Queue {} does not exist, scheduling re-subscription", queueName);
+                        this.topologyRecoveryScheduledExecutorService.schedule(
+                            resubscriptionReference.get(), schedulingPeriod.getSeconds(), TimeUnit.SECONDS
+                        );
+                    } else {
+                       LOGGER.debug("Max subscription retry count reached {} for queue {}",
+                           retryCount.get(), queueName);
+                    }
+                    return null;
+                };
+                resubscriptionReference.set(resubscription);
+                this.topologyRecoveryScheduledExecutorService.schedule(
+                    resubscription, schedulingPeriod.getSeconds(), TimeUnit.SECONDS
+                );
+            }
+        } else {
+            channel.basicConsume(queueName, autoAck, consumerTag, false, false, this.consumerArguments, q);
+        }
+    }
+
+    private static boolean queueMayBeDown(RecordedQueue queueRecord, TopologyRecording topologyRecording) {
+        return queueRecord != null
+            && queueRecord.isClassic() && queueRecord.isDurable() && topologyRecording.isCluster();
     }
 
     void maybeStopIfNoActivityOrQueueEmpty() {

--- a/src/main/java/com/rabbitmq/perf/ConsumerParameters.java
+++ b/src/main/java/com/rabbitmq/perf/ConsumerParameters.java
@@ -1,4 +1,4 @@
-// Copyright (c) 2019-2020 VMware, Inc. or its affiliates.  All rights reserved.
+// Copyright (c) 2019-2022 VMware, Inc. or its affiliates.  All rights reserved.
 //
 // This software, the RabbitMQ Java client library, is triple-licensed under the
 // Mozilla Public License 2.0 ("MPL"), the GNU General Public License version 2
@@ -21,6 +21,7 @@ import com.rabbitmq.perf.PerfTest.EXIT_WHEN;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
 
 /**
  *
@@ -51,6 +52,8 @@ public class ConsumerParameters {
     private EXIT_WHEN exitWhen = EXIT_WHEN.NEVER;
 
     private Map<String, Object> consumerArguments = null;
+
+    private ScheduledExecutorService topologyRecoveryScheduledExecutorService;
 
     public Channel getChannel() {
         return channel;
@@ -230,5 +233,15 @@ public class ConsumerParameters {
 
     public EXIT_WHEN getExitWhen() {
         return exitWhen;
+    }
+
+    ConsumerParameters setTopologyRecoveryScheduledExecutorService(
+        ScheduledExecutorService topologyRecoveryScheduledExecutorService) {
+        this.topologyRecoveryScheduledExecutorService = topologyRecoveryScheduledExecutorService;
+        return this;
+    }
+
+    public ScheduledExecutorService getTopologyRecoveryScheduledExecutorService() {
+        return topologyRecoveryScheduledExecutorService;
     }
 }

--- a/src/main/java/com/rabbitmq/perf/MulticastParams.java
+++ b/src/main/java/com/rabbitmq/perf/MulticastParams.java
@@ -18,7 +18,6 @@ package com.rabbitmq.perf;
 import com.rabbitmq.client.AMQP;
 import com.rabbitmq.client.Channel;
 import com.rabbitmq.client.Connection;
-import com.rabbitmq.client.ShutdownSignalException;
 
 import com.rabbitmq.perf.PerfTest.EXIT_WHEN;
 import java.io.IOException;
@@ -31,11 +30,13 @@ import java.util.UUID;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeoutException;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.Collectors;
 
 import static com.rabbitmq.perf.Recovery.setupRecoveryProcess;
+import static com.rabbitmq.perf.Utils.exists;
 
 public class MulticastParams {
 
@@ -126,6 +127,8 @@ public class MulticastParams {
 
     // for random JSON body generation
     private AtomicReference<MessageBodySource> messageBodySourceReference = new AtomicReference<>();
+
+    private boolean cluster = false;
 
     public void setExchangeType(String exchangeType) {
         this.exchangeType = exchangeType;
@@ -291,6 +294,10 @@ public class MulticastParams {
 
     public void setExitWhen(EXIT_WHEN exitWhen) {
         this.exitWhen = exitWhen;
+    }
+
+    void setCluster(boolean cluster) {
+        this.cluster = cluster;
     }
 
     public int getConsumerCount() {
@@ -475,7 +482,7 @@ public class MulticastParams {
         Channel channel = connection.createChannel(); //NOSONAR
         if (producerTxSize > 0) channel.txSelect();
         if (confirm >= 0) channel.confirmSelect();
-        TopologyRecording topologyRecording = new TopologyRecording(this.isPolling());
+        TopologyRecording topologyRecording = new TopologyRecording(this.isPolling(), this.cluster);
         if (!predeclared || !exchangeExists(connection, exchangeName)) {
             Utils.exchangeDeclare(channel, exchangeName, exchangeType);
             topologyRecording.recordExchange(exchangeName, exchangeType);
@@ -519,7 +526,8 @@ public class MulticastParams {
                                    Stats stats,
                                    ValueIndicator<Long> consumerLatenciesIndicator,
                                    MulticastSet.CompletionHandler completionHandler,
-                                   ExecutorService executorService) throws IOException {
+                                   ExecutorService executorService,
+                                   ScheduledExecutorService topologyRecordingScheduledExecutorService) throws IOException {
         TopologyHandlerResult topologyHandlerResult = this.topologyHandler.configureQueuesForClient(connection);
         connection = topologyHandlerResult.connection;
         Channel channel = connection.createChannel(); //NOSONAR
@@ -559,17 +567,20 @@ public class MulticastParams {
                 .setRequeue(this.requeue)
                 .setConsumerArguments(this.consumerArguments)
                 .setExitWhen(this.exitWhen)
+                .setTopologyRecoveryScheduledExecutorService(topologyRecordingScheduledExecutorService)
         );
         this.topologyHandler.next();
         return consumer;
     }
 
-    public List<TopologyHandlerResult> configureAllQueues(List<Connection> connection) throws IOException {
-        return this.topologyHandler.configureAllQueues(connection);
+    public List<TopologyHandlerResult> configureAllQueues(List<Connection> connections) throws IOException {
+        return this.topologyHandler.configureAllQueues(connections);
     }
 
     public void init() {
-        this.topologyRecording = new TopologyRecording(this.isPolling());
+        this.topologyRecording = new TopologyRecording(
+            this.isPolling(), this.cluster
+        );
         if (this.queuePattern == null && !this.queuesInSequence) {
             this.topologyHandler = new FixedQueuesTopologyHandler(this, this.routingKey, this.queueNames, topologyRecording);
         } else if (this.queuePattern == null && this.queuesInSequence) {
@@ -648,29 +659,6 @@ public class MulticastParams {
 
     public void setProducerSchedulerThreadCount(int producerSchedulerThreadCount) {
         this.producerSchedulerThreadCount = producerSchedulerThreadCount;
-    }
-
-    private interface Checker {
-        void check(Channel ch) throws IOException;
-    }
-
-    private static boolean exists(Connection connection, Checker checker) throws IOException {
-        try {
-            Channel ch = connection.createChannel();
-            checker.check(ch);
-            ch.abort();
-            return true;
-        }
-        catch (IOException e) {
-            ShutdownSignalException sse = (ShutdownSignalException) e.getCause();
-            if (!sse.isHardError()) {
-                AMQP.Channel.Close closeMethod = (AMQP.Channel.Close) sse.getReason();
-                if (closeMethod.getReplyCode() == AMQP.NOT_FOUND) {
-                    return false;
-                }
-            }
-            throw e;
-        }
     }
 
     /**

--- a/src/main/java/com/rabbitmq/perf/MulticastSet.java
+++ b/src/main/java/com/rabbitmq/perf/MulticastSet.java
@@ -208,6 +208,13 @@ public class MulticastSet {
             List<MulticastParams.TopologyHandlerResult> topologyHandlerResults = params.configureAllQueues(configurationConnections);
             enableTopologyRecoveryIfNecessary(topologyHandlerResults);
 
+            ScheduledExecutorService topologyRecordingScheduledExecutorService = null;
+            if (!configurationConnections.isEmpty() && Utils.isRecoverable(configurationConnections.get(0))) {
+                topologyRecordingScheduledExecutorService = this.threadingHandler.scheduledExecutorService(
+                  "perf-test-topology-recovery-", 1
+                );
+            }
+
             this.params.resetTopologyHandler();
 
             Consumer[] consumerRunnables = new Consumer[params.getConsumerThreadCount()];
@@ -215,7 +222,8 @@ public class MulticastSet {
             Function<Integer, ExecutorService> consumersExecutorsFactory;
             consumersExecutorsFactory = createConsumersExecutorsFactory();
 
-            createConsumers(announceStartup, consumerRunnables, consumerConnections, consumersExecutorsFactory);
+            createConsumers(announceStartup, consumerRunnables, consumerConnections,
+                consumersExecutorsFactory, topologyRecordingScheduledExecutorService);
 
             this.params.resetTopologyHandler();
 
@@ -376,7 +384,8 @@ public class MulticastSet {
     private void createConsumers(boolean announceStartup,
                                  Runnable[] consumerRunnables,
                                  Connection[] consumerConnections,
-                                 Function<Integer, ExecutorService> consumersExecutorsFactory) throws IOException, TimeoutException {
+                                 Function<Integer, ExecutorService> consumersExecutorsFactory,
+                                 ScheduledExecutorService topologyRecordingScheduledExecutorService) throws IOException, TimeoutException {
         for (int i = 0; i < consumerConnections.length; i++) {
             if (announceStartup) {
                 System.out.println("id: " + testID + ", starting consumer #" + i);
@@ -390,13 +399,16 @@ public class MulticastSet {
                 if (announceStartup) {
                     System.out.println("id: " + testID + ", starting consumer #" + i + ", channel #" + j);
                 }
-                Consumer consumer = params.createConsumer(consumerConnection, stats, this.consumerLatencyIndicator, this.completionHandler, executorService);
+                Consumer consumer = params.createConsumer(consumerConnection, stats,
+                    this.consumerLatencyIndicator, this.completionHandler, executorService,
+                    topologyRecordingScheduledExecutorService);
                 consumerRunnables[(i * params.getConsumerChannelCount()) + j] = consumer;
             }
         }
     }
 
-    private void createProducers(boolean announceStartup, AgentState[] producerStates, Connection[] producerConnections) throws IOException, TimeoutException {
+    private void createProducers(boolean announceStartup, AgentState[] producerStates,
+                                 Connection[] producerConnections) throws IOException, TimeoutException {
         for (int i = 0; i < producerConnections.length; i++) {
             if (announceStartup) {
                 System.out.println("id: " + testID + ", starting producer #" + i);

--- a/src/main/java/com/rabbitmq/perf/PerfTest.java
+++ b/src/main/java/com/rabbitmq/perf/PerfTest.java
@@ -406,6 +406,7 @@ public class PerfTest {
             p.setConsumerArguments(convertKeyValuePairs(consumerArgs));
             p.setQueuesInSequence(queueFile != null);
             p.setExitWhen(exitWhen);
+            p.setCluster(uris.size() > 0);
 
             ConcurrentMap<String, Integer> completionReasons = new ConcurrentHashMap<>();
 

--- a/src/main/java/com/rabbitmq/perf/Recovery.java
+++ b/src/main/java/com/rabbitmq/perf/Recovery.java
@@ -1,4 +1,4 @@
-// Copyright (c) 2007-2020 VMware, Inc. or its affiliates.  All rights reserved.
+// Copyright (c) 2007-2022 VMware, Inc. or its affiliates.  All rights reserved.
 //
 // This software, the RabbitMQ Java client library, is triple-licensed under the
 // Mozilla Public License 2.0 ("MPL"), the GNU General Public License version 2
@@ -57,6 +57,7 @@ public class Recovery {
                 @Override
                 public void init(AgentBase agent) {
                     agentReference.set(agent);
+                    agent.setTopologyRecording(topologyRecording);
                 }
 
                 @Override
@@ -74,6 +75,7 @@ public class Recovery {
 
                 @Override
                 public void handleRecoveryStarted(Recoverable recoverable) {
+                    // FIXME use lock to avoid start topology recovery twice
                     recoveryInProgress.set(true);
                 }
 

--- a/src/main/java/com/rabbitmq/perf/TopologyRecording.java
+++ b/src/main/java/com/rabbitmq/perf/TopologyRecording.java
@@ -17,6 +17,8 @@ package com.rabbitmq.perf;
 
 import com.rabbitmq.client.Channel;
 import com.rabbitmq.client.Connection;
+import java.util.Collections;
+import java.util.function.Supplier;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -42,6 +44,7 @@ public class TopologyRecording {
     private final Collection<TopologyRecording> children = new CopyOnWriteArrayList<>();
 
     private final boolean polling;
+    private final boolean cluster;
 
     /**
      * Create  a {@link TopologyRecording} with or without polling option.
@@ -52,16 +55,25 @@ public class TopologyRecording {
      *
      * @param polling
      */
-    public TopologyRecording(boolean polling) {
+    public TopologyRecording(boolean polling, boolean cluster) {
         this.polling = polling;
+        this.cluster = cluster;
     }
 
     private static Channel reliableWrite(Connection connection, Channel channel, WriteOperation operation) throws IOException {
+        return reliableWrite(connection, channel, operation, () -> "");
+    }
+
+    private static Channel reliableWrite(Connection connection, Channel channel, WriteOperation operation, Supplier<String> message) throws IOException {
         try {
             operation.write(channel);
             return channel;
         } catch (Exception e) {
             LOGGER.warn("Error during topology recovery: {}", e.getMessage());
+            String msg = message.get();
+            if (msg != null && !msg.isEmpty()) {
+               LOGGER.debug(msg);
+            }
             return connection.createChannel();
         }
     }
@@ -76,7 +88,7 @@ public class TopologyRecording {
      * @return
      */
     TopologyRecording child() {
-        TopologyRecording child = new TopologyRecording(this.polling);
+        TopologyRecording child = new TopologyRecording(this.polling, this.cluster);
         this.children.add(child);
         return child;
     }
@@ -176,16 +188,18 @@ public class TopologyRecording {
                         }
                     }
                     if (redeclare) {
+                        LOGGER.debug("Trying to re-declare queue {}", originalName);
                         channel = reliableWrite(connection, channel, ch -> {
                             String newName = ch.queueDeclare(
                                     queue.serverNamed ? "" : queue.name, queue.durable, queue.exclusive,
                                     queue.autoDelete, queue.arguments
                             ).getQueue();
+                            LOGGER.debug("Re-declared queue {}", originalName);
                             queue.name = newName;
                             if (queue.serverNamed) {
                                 LOGGER.debug("Queue {} was server-named, it is now {}", originalName, newName);
                             }
-                        });
+                        }, () -> "Error while trying to re-declare queue " + originalName);
                     }
 
                     LOGGER.debug("Connection {}, recovered queue {}", connection.getClientProvidedName(), queue);
@@ -237,6 +251,10 @@ public class TopologyRecording {
         void write(Channel channel) throws IOException;
     }
 
+    boolean isCluster() {
+        return this.cluster;
+    }
+
     class RecordedExchange {
 
         private final String name, type;
@@ -264,12 +282,12 @@ public class TopologyRecording {
         private final boolean serverNamed;
         private String name;
 
-        public RecordedQueue(String name, boolean durable, boolean exclusive, boolean autoDelete, Map<String, Object> arguments, boolean serverNamed) {
+        private RecordedQueue(String name, boolean durable, boolean exclusive, boolean autoDelete, Map<String, Object> arguments, boolean serverNamed) {
             this.name = name;
             this.durable = durable;
             this.exclusive = exclusive;
             this.autoDelete = autoDelete;
-            this.arguments = arguments;
+            this.arguments = arguments == null ? Collections.emptyMap() : arguments;
             this.serverNamed = serverNamed;
         }
 
@@ -287,6 +305,22 @@ public class TopologyRecording {
 
         public boolean isExclusive() {
             return exclusive;
+        }
+
+        boolean isQuorum() {
+            return "quorum".equals(this.arguments.get("x-queue-type"));
+        }
+
+        boolean isStream() {
+            return "stream".equals(this.arguments.get("x-queue-type"));
+        }
+
+        boolean isClassic() {
+            return !isQuorum() && !isStream();
+        }
+
+        boolean isDurable() {
+            return durable;
         }
 
         @Override

--- a/src/main/java/com/rabbitmq/perf/Utils.java
+++ b/src/main/java/com/rabbitmq/perf/Utils.java
@@ -1,4 +1,4 @@
-// Copyright (c) 2018-2020 VMware, Inc. or its affiliates.  All rights reserved.
+// Copyright (c) 2018-2022 VMware, Inc. or its affiliates.  All rights reserved.
 //
 // This software, the RabbitMQ Java client library, is triple-licensed under the
 // Mozilla Public License 2.0 ("MPL"), the GNU General Public License version 2
@@ -15,11 +15,13 @@
 
 package com.rabbitmq.perf;
 
+import com.rabbitmq.client.AMQP;
 import com.rabbitmq.client.Address;
 import com.rabbitmq.client.Channel;
 import com.rabbitmq.client.Connection;
 import com.rabbitmq.client.ConnectionFactory;
 import com.rabbitmq.client.RecoveryDelayHandler;
+import com.rabbitmq.client.ShutdownSignalException;
 import com.rabbitmq.client.SocketConfigurator;
 import com.rabbitmq.client.SocketConfigurators;
 import com.rabbitmq.client.SslEngineConfigurator;
@@ -188,5 +190,28 @@ abstract class Utils {
       LOGGER.info("Skipping creation of exchange {}", exchange);
     }
     channel.exchangeDeclare(exchange, type);
+  }
+
+  interface Checker {
+    void check(Channel ch) throws IOException;
+  }
+
+  static boolean exists(Connection connection, Checker checker) throws IOException {
+    try {
+      Channel ch = connection.createChannel();
+      checker.check(ch);
+      ch.abort();
+      return true;
+    }
+    catch (IOException e) {
+      ShutdownSignalException sse = (ShutdownSignalException) e.getCause();
+      if (!sse.isHardError()) {
+        AMQP.Channel.Close closeMethod = (AMQP.Channel.Close) sse.getReason();
+        if (closeMethod.getReplyCode() == AMQP.NOT_FOUND) {
+          return false;
+        }
+      }
+      throw e;
+    }
   }
 }

--- a/src/test/java/com/rabbitmq/perf/ConnectionRecoveryTest.java
+++ b/src/test/java/com/rabbitmq/perf/ConnectionRecoveryTest.java
@@ -1,0 +1,229 @@
+// Copyright (c) 2022 VMware, Inc. or its affiliates.  All rights reserved.
+//
+// This software, the RabbitMQ Java client library, is triple-licensed under the
+// Mozilla Public License 2.0 ("MPL"), the GNU General Public License version 2
+// ("GPL") and the Apache License version 2 ("ASL"). For the MPL, please see
+// LICENSE-MPL-RabbitMQ. For the GPL, please see LICENSE-GPL2.  For the ASL,
+// please see LICENSE-APACHE2.
+//
+// This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY KIND,
+// either express or implied. See the LICENSE file for specific language governing
+// rights and limitations of this software.
+//
+// If you have any questions regarding licensing, please contact us at
+// info@rabbitmq.com.
+
+package com.rabbitmq.perf;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyBoolean;
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.ArgumentMatchers.anyList;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.ArgumentMatchers.anyMap;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.ArgumentMatchers.isNull;
+import static org.mockito.ArgumentMatchers.nullable;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.mockito.MockitoAnnotations.openMocks;
+
+import com.rabbitmq.client.AMQP;
+import com.rabbitmq.client.AMQP.Queue.DeclareOk;
+import com.rabbitmq.client.Channel;
+import com.rabbitmq.client.ConnectionFactory;
+import com.rabbitmq.client.Consumer;
+import com.rabbitmq.client.RecoveryListener;
+import com.rabbitmq.client.ShutdownListener;
+import com.rabbitmq.client.ShutdownSignalException;
+import com.rabbitmq.client.impl.AMQImpl;
+import com.rabbitmq.client.impl.AMQImpl.Channel.Close;
+import com.rabbitmq.client.impl.recovery.AutorecoveringConnection;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mock;
+
+/** */
+public class ConnectionRecoveryTest {
+
+  private static final String QUEUE = "test-queue";
+  @Mock ConnectionFactory cf;
+  @Mock AutorecoveringConnection c;
+  @Mock Channel ch;
+  @Mock Stats stats;
+  @Mock MulticastSet.ThreadingHandler threadingHandler;
+  @Mock ExecutorService executorService;
+  @Mock ScheduledExecutorService scheduledExecutorService;
+  @Mock Future future;
+  MulticastParams params;
+  AutoCloseable mocks;
+
+  @BeforeEach
+  @SuppressWarnings("unchecked")
+  public void init() throws Exception {
+    mocks = openMocks(this);
+
+    when(cf.newConnection(anyList(), anyString())).thenReturn(c);
+    when(c.createChannel()).thenReturn(ch);
+    when(c.getClientProvidedName()).thenReturn("consumer-connection");
+    DeclareOk declareOk = new AMQImpl.Queue.DeclareOk(QUEUE, 0, 0);
+    when(ch.queueDeclare(eq(QUEUE), anyBoolean(), anyBoolean(), anyBoolean(), anyMap()))
+        .thenReturn(declareOk);
+    when(ch.queueDeclare(eq(QUEUE), anyBoolean(), anyBoolean(), anyBoolean(), isNull()))
+        .thenReturn(declareOk);
+    when(ch.getConnection()).thenReturn(c);
+    when(ch.queueDeclarePassive(QUEUE))
+        .thenThrow(
+            new IOException(
+                new ShutdownSignalException(
+                    false, false, new Close(AMQP.NOT_FOUND, "", 0, 0), null)))
+        .thenReturn(declareOk);
+
+    when(threadingHandler.executorService(anyString(), anyInt())).thenReturn(executorService);
+    when(threadingHandler.scheduledExecutorService(anyString(), anyInt()))
+        .thenReturn(scheduledExecutorService);
+    when(executorService.submit(any(Runnable.class))).thenReturn(future);
+
+    params = new MulticastParams();
+    params.setQueueNames(Collections.singletonList(QUEUE));
+    params.setAutoDelete(false);
+    params.setExclusive(false);
+    params.setFlags(Collections.singletonList("persistent"));
+    params.setCluster(true);
+  }
+
+  @AfterEach
+  public void tearDown() throws Exception {
+    mocks.close();
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  void subscriptionShouldBeScheduledIfQueueDoesNotExist() throws Exception {
+    // for https://github.com/rabbitmq/rabbitmq-perf-test/issues/330
+    List<RecoveryListener> recoveryListeners = Collections.synchronizedList(new ArrayList<>());
+    doAnswer(
+            invocation -> {
+              recoveryListeners.add(invocation.getArgument(0, RecoveryListener.class));
+              return null;
+            })
+        .when(c)
+        .addRecoveryListener(any(RecoveryListener.class));
+    List<ShutdownListener> shutdownListeners = Collections.synchronizedList(new ArrayList<>());
+    doAnswer(
+            invocation -> {
+              shutdownListeners.add(invocation.getArgument(0, ShutdownListener.class));
+              return null;
+            })
+        .when(c)
+        .addShutdownListener(any(ShutdownListener.class));
+    CountDownLatch completionLatch = new CountDownLatch(1);
+    CountDownLatch runStartedLatch = new CountDownLatch(1);
+    new Thread(
+            () -> {
+              try {
+                MulticastSet set = getMulticastSet(runStartedLatch, completionLatch);
+                set.run();
+              } catch (Exception e) {
+                throw new RuntimeException(e);
+              }
+            })
+        .start();
+
+    assertThat(runStartedLatch.await(10, TimeUnit.SECONDS)).isTrue();
+
+    assertThat(recoveryListeners)
+        .hasSize(5); // configuration x 3 (cluster) + consumer + producer connections
+    assertThat(shutdownListeners).hasSize(2); // consumer + producer connections
+
+    // initial subscription
+    verify(ch, times(1))
+        .basicConsume(anyString(), anyBoolean(), nullable(Map.class), any(Consumer.class));
+    // subscription after recovery, not called yet
+    verify(ch, never())
+        .basicConsume(
+            anyString(),
+            anyBoolean(),
+            anyString(),
+            anyBoolean(),
+            anyBoolean(),
+            nullable(Map.class),
+            any(Consumer.class));
+
+    // synchronously call the scheduled subscription
+    doAnswer(
+            invocation -> {
+              Callable<?> action = invocation.getArgument(0);
+              action.call();
+              return null;
+            })
+        .when(scheduledExecutorService)
+        .schedule(any(Callable.class), anyLong(), any(TimeUnit.class));
+
+    RecoveryListener consumerRecoveryListener = recoveryListeners.get(3);
+    ShutdownListener consumerShutdownListener = shutdownListeners.get(1);
+
+    // mimic connection recovery
+    consumerShutdownListener.shutdownCompleted(
+        new ShutdownSignalException(false, true, null, null));
+    consumerRecoveryListener.handleRecoveryStarted(null);
+    consumerRecoveryListener.handleRecovery(null);
+
+    // first check fails, second check says the queue exists
+    verify(ch, times(2)).queueDeclarePassive(QUEUE);
+    // initial subscription, still the same
+    verify(ch, times(1))
+        .basicConsume(anyString(), anyBoolean(), nullable(Map.class), any(Consumer.class));
+    // subscription after recovery
+    verify(ch, times(1))
+        .basicConsume(
+            anyString(),
+            anyBoolean(),
+            nullable(String.class),
+            anyBoolean(),
+            anyBoolean(),
+            nullable(Map.class),
+            any(Consumer.class));
+  }
+
+  private MulticastSet getMulticastSet(
+      CountDownLatch runStartedLatch, CountDownLatch completionLatch) {
+    MulticastSet set =
+        new MulticastSet(
+            stats,
+            cf,
+            params,
+            Arrays.asList("amqp://localhost", "amqp://localhost", "amqp://localhost"),
+            new MulticastSet.CompletionHandler() {
+
+              @Override
+              public void waitForCompletion() throws InterruptedException {
+                runStartedLatch.countDown();
+                completionLatch.await(10, TimeUnit.SECONDS);
+              }
+
+              @Override
+              public void countDown(String reason) {}
+            });
+    set.setThreadingHandler(threadingHandler);
+    return set;
+  }
+}

--- a/src/test/java/com/rabbitmq/perf/SequenceTopologyHandlerTest.java
+++ b/src/test/java/com/rabbitmq/perf/SequenceTopologyHandlerTest.java
@@ -1,4 +1,4 @@
-// Copyright (c) 2018-2020 VMware, Inc. or its affiliates.  All rights reserved.
+// Copyright (c) 2018-2022 VMware, Inc. or its affiliates.  All rights reserved.
 //
 // This software, the RabbitMQ Java client library, is triple-licensed under the
 // Mozilla Public License 2.0 ("MPL"), the GNU General Public License version 2
@@ -26,7 +26,7 @@ public class SequenceTopologyHandlerTest {
 
     @Test
     public void sequence() {
-        handler = new MulticastParams.SequenceTopologyHandler(null, 1, 5, "test-%d", new TopologyRecording(false), null);
+        handler = new MulticastParams.SequenceTopologyHandler(null, 1, 5, "test-%d", new TopologyRecording(false, false), null);
         assertThat(handler.getQueueNames()).hasSize(5).contains("test-1", "test-2", "test-3", "test-4", "test-5");
 
         assertThat(handler.getRoutingKey()).isEqualTo("test-1");
@@ -51,7 +51,7 @@ public class SequenceTopologyHandlerTest {
 
     @Test
     public void useFixedRoutingKeyWhenProvided() {
-        handler = new MulticastParams.SequenceTopologyHandler(null, 1, 5, "test-%d", new TopologyRecording(false), "rk");
+        handler = new MulticastParams.SequenceTopologyHandler(null, 1, 5, "test-%d", new TopologyRecording(false, false), "rk");
         assertThat(handler.getQueueNames()).hasSize(5).contains("test-1", "test-2", "test-3", "test-4", "test-5");
 
         assertThat(handler.getQueueNamesForClient()).hasSize(1).contains("test-1");
@@ -63,7 +63,7 @@ public class SequenceTopologyHandlerTest {
 
     @Test
     public void reset() {
-        handler = new MulticastParams.SequenceTopologyHandler(null, 1, 100, "test-%d", new TopologyRecording(false), null);
+        handler = new MulticastParams.SequenceTopologyHandler(null, 1, 100, "test-%d", new TopologyRecording(false, false), null);
         assertThat(handler.getQueueNames()).hasSize(100);
 
         assertThat(handler.getRoutingKey()).isEqualTo("test-1");
@@ -86,7 +86,7 @@ public class SequenceTopologyHandlerTest {
 
     @Test
     public void format() {
-        handler = new MulticastParams.SequenceTopologyHandler(null, 1, 5, "test-%03d", new TopologyRecording(false), null);
+        handler = new MulticastParams.SequenceTopologyHandler(null, 1, 5, "test-%03d", new TopologyRecording(false, false), null);
         assertThat(handler.getQueueNames()).hasSize(5).contains("test-001", "test-002", "test-003", "test-004", "test-005");
     }
 }

